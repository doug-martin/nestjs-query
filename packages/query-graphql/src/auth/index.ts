export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
