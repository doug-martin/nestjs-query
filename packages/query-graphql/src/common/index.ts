export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
