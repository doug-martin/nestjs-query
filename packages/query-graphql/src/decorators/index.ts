export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
