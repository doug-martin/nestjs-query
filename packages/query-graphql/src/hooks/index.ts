export * from './hooks';
export { createDefaultHook } from './default.hook';
export * from './tokens';
export * from './types';
