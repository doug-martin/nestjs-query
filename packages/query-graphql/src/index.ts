export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
