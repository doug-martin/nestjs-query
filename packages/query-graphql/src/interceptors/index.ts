export * from './hook.interceptor';
export * from './authorizer.interceptor';
