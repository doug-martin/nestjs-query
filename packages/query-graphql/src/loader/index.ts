export * from './find-relations.loader';
export * from './query-relations.loader';
export * from './count-relations.loader';
export * from './aggregate-relations.loader';
export * from './dataloader.factory';
