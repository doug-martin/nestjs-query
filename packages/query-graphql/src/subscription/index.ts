import { Class } from '@nestjs-query/core';
import { PubSub } from 'graphql-subscriptions';
import { DTONamesOpts, getDTONames } from '../common';
import { GraphQLPubSub } from './pub-sub.interface';

export { GraphQLPubSub } from './pub-sub.interface';

export enum EventType {
  CREATED = 'created',
  UPDATED_ONE = 'updatedOne',
  UPDATED_MANY = 'updatedMany',
  DELETED_ONE = 'deletedOne',
  DELETED_MANY = 'deletedMany',
}
export const getDTOEventName = <DTO>(type: EventType, DTOClass: Class<DTO>, opts?: DTONamesOpts): string => {
  const { baseName, pluralBaseName } = getDTONames(DTOClass, opts);
  if (type === EventType.DELETED_MANY || type === EventType.UPDATED_MANY) {
    return `${type}${pluralBaseName}`;
  }
  return `${type}${baseName}`;
};

export const pubSubToken = (): string => 'pub_sub';

let pubSub: GraphQLPubSub;
export const defaultPubSub = (): GraphQLPubSub => {
  if (!pubSub) {
    pubSub = new PubSub();
  }
  return pubSub;
};
