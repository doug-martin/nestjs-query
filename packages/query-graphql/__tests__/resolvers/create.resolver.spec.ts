// eslint-disable-next-line max-classes-per-file
import { DeepPartial } from '@nestjs-query/core';
import { Resolver, Query, InputType } from '@nestjs/graphql';
import { when, objectContaining, verify, deepEqual, anything, mock, instance } from 'ts-mockito';
import { PubSub } from 'graphql-subscriptions';
import { CreateManyInputType, CreateOneInputType, CreateResolver, CreateResolverOpts, InjectPubSub } from '../../src';
import { CreatedEvent } from '../../src/resolvers/create.resolver';
import { EventType, getDTOEventName } from '../../src/subscription';
import {
  generateSchema,
  createResolverFromNest,
  TestResolverDTO,
  TestResolverInputDTO,
  TestService,
} from '../__fixtures__';

describe('CreateResolver', () => {
  const expectResolverSDL = async (opts?: CreateResolverOpts<TestResolverDTO>) => {
    @Resolver(() => TestResolverDTO)
    class TestSDLResolver extends CreateResolver(TestResolverDTO, opts) {
      @Query(() => TestResolverDTO)
      test(): TestResolverDTO {
        return { id: '1', stringField: 'foo' };
      }
    }

    const schema = await generateSchema([TestSDLResolver]);
    expect(schema).toMatchSnapshot();
  };

  const createTestResolver = (opts?: CreateResolverOpts<TestResolverDTO>) => {
    @Resolver(() => TestResolverDTO)
    class TestResolver extends CreateResolver(TestResolverDTO, opts) {
      constructor(service: TestService, @InjectPubSub() readonly pubSub: PubSub) {
        super(service);
      }
    }

    return createResolverFromNest(TestResolver);
  };

  it('should create a CreateResolver for the DTO', () => expectResolverSDL());

  it('should use the dtoName if provided', () => expectResolverSDL({ dtoName: 'Test' }));

  it('should use the one.name option for the createOne if provided', () =>
    expectResolverSDL({ one: { name: 'create_one_test' } }));

  it('should use the many.name option for the createMany if provided', () =>
    expectResolverSDL({ many: { name: 'create_many_test' } }));

  it('should use the CreateDTOClass if provided', () => expectResolverSDL({ CreateDTOClass: TestResolverInputDTO }));

  it('should not expose create methods if disabled', () => expectResolverSDL({ disabled: true }));

  describe('#createOne', () => {
    it('should use the provided CreateOneInput type', () => {
      @InputType()
      class CreateOneInput extends CreateOneInputType('createResolverDTO', TestResolverInputDTO) {}

      return expectResolverSDL({ CreateOneInput });
    });

    it('should not expose create one method if disabled', () => expectResolverSDL({ one: { disabled: true } }));

    it('should call the service createOne with the provided input', async () => {
      const { resolver, mockService } = await createTestResolver();
      const args: CreateOneInputType<DeepPartial<TestResolverDTO>> = {
        input: {
          stringField: 'foo',
        },
      };
      const output: TestResolverDTO = {
        id: 'id-1',
        stringField: 'foo',
      };
      when(mockService.createOne(objectContaining(args.input))).thenResolve(output);
      const result = await resolver.createOne({ input: args });
      return expect(result).toEqual(output);
    });
  });

  describe('#createMany', () => {
    it('should not create a new type if the CreateManyArgs is supplied', () => {
      @InputType()
      class CreateManyInput extends CreateManyInputType('testResolvers', TestResolverInputDTO) {}

      return expectResolverSDL({ CreateManyInput });
    });

    it('should not expose create many method if disabled', () => expectResolverSDL({ many: { disabled: true } }));

    it('should call the service createMany with the provided input', async () => {
      const { resolver, mockService } = await createTestResolver();
      const args: CreateManyInputType<Partial<TestResolverDTO>> = {
        input: [
          {
            stringField: 'foo',
          },
        ],
      };
      const output: TestResolverDTO[] = [
        {
          id: 'id-1',
          stringField: 'foo',
        },
      ];
      when(mockService.createMany(objectContaining(args.input))).thenResolve(output);
      const result = await resolver.createMany({ input: args });
      return expect(result).toEqual(output);
    });
  });

  describe('created subscription', () => {
    it('should add subscription types if enableSubscriptions is true', () =>
      expectResolverSDL({ enableSubscriptions: true }));

    it('should not expose subscriptions if enableSubscriptions is false', () =>
      expectResolverSDL({ enableSubscriptions: false }));

    describe('create one events', () => {
      it('should publish events for create one when enableSubscriptions is set to true for all', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({ enableSubscriptions: true });
        const args: CreateOneInputType<DeepPartial<TestResolverDTO>> = {
          input: {
            stringField: 'foo',
          },
        };
        const output: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const eventName = getDTOEventName(EventType.CREATED, TestResolverDTO);
        const event = { [eventName]: output };
        when(mockService.createOne(objectContaining(args.input))).thenResolve(output);
        when(mockPubSub.publish(eventName, deepEqual(event))).thenResolve();
        const result = await resolver.createOne({ input: args });
        verify(mockPubSub.publish(eventName, deepEqual(event))).once();
        return expect(result).toEqual(output);
      });

      it('should publish events for create one when enableSubscriptions is set to true for createOne', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({ one: { enableSubscriptions: true } });
        const args: CreateOneInputType<DeepPartial<TestResolverDTO>> = {
          input: {
            stringField: 'foo',
          },
        };
        const output: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const eventName = getDTOEventName(EventType.CREATED, TestResolverDTO);
        const event = { [eventName]: output };
        when(mockService.createOne(objectContaining(args.input))).thenResolve(output);
        when(mockPubSub.publish(eventName, deepEqual(event))).thenResolve();
        const result = await resolver.createOne({ input: args });
        verify(mockPubSub.publish(eventName, deepEqual(event))).once();
        return expect(result).toEqual(output);
      });

      it('should not publish an event if enableSubscriptions is false', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({ enableSubscriptions: false });
        const args: CreateOneInputType<DeepPartial<TestResolverDTO>> = {
          input: {
            stringField: 'foo',
          },
        };
        const output: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        when(mockService.createOne(objectContaining(args.input))).thenResolve(output);
        const result = await resolver.createOne({ input: args });
        verify(mockPubSub.publish(anything(), anything())).never();
        return expect(result).toEqual(output);
      });

      it('should not publish an event if enableSubscriptions is true and one.enableSubscriptions is false', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({
          enableSubscriptions: true,
          one: { enableSubscriptions: false },
        });
        const args: CreateOneInputType<DeepPartial<TestResolverDTO>> = {
          input: {
            stringField: 'foo',
          },
        };
        const output: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        when(mockService.createOne(objectContaining(args.input))).thenResolve(output);
        const result = await resolver.createOne({ input: args });
        verify(mockPubSub.publish(anything(), anything())).never();
        return expect(result).toEqual(output);
      });
    });

    describe('create many events', () => {
      it('should publish events for create many when enableSubscriptions is set to true for all', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({ enableSubscriptions: true });
        const args: CreateManyInputType<Partial<TestResolverDTO>> = {
          input: [
            {
              stringField: 'foo',
            },
          ],
        };
        const output: TestResolverDTO[] = [
          {
            id: 'id-1',
            stringField: 'foo',
          },
        ];
        const eventName = getDTOEventName(EventType.CREATED, TestResolverDTO);
        const events = output.map((o) => ({ [eventName]: o }));
        when(mockService.createMany(objectContaining(args.input))).thenResolve(output);
        events.forEach((e) => when(mockPubSub.publish(eventName, deepEqual(e))).thenResolve());
        const result = await resolver.createMany({ input: args });
        events.forEach((e) => verify(mockPubSub.publish(eventName, deepEqual(e))).once());
        return expect(result).toEqual(output);
      });

      it('should publish events for create one when enableSubscriptions is set to true for createOne', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({ many: { enableSubscriptions: true } });
        const args: CreateManyInputType<Partial<TestResolverDTO>> = {
          input: [
            {
              stringField: 'foo',
            },
          ],
        };
        const output: TestResolverDTO[] = [
          {
            id: 'id-1',
            stringField: 'foo',
          },
        ];
        const eventName = getDTOEventName(EventType.CREATED, TestResolverDTO);
        const events = output.map((o) => ({ [eventName]: o }));
        when(mockService.createMany(objectContaining(args.input))).thenResolve(output);
        events.forEach((e) => when(mockPubSub.publish(eventName, deepEqual(e))).thenResolve());
        const result = await resolver.createMany({ input: args });
        events.forEach((e) => verify(mockPubSub.publish(eventName, deepEqual(e))).once());
        return expect(result).toEqual(output);
      });

      it('should not publish an event if enableSubscriptions is false', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({ enableSubscriptions: false });
        const args: CreateManyInputType<Partial<TestResolverDTO>> = {
          input: [
            {
              stringField: 'foo',
            },
          ],
        };
        const output: TestResolverDTO[] = [
          {
            id: 'id-1',
            stringField: 'foo',
          },
        ];
        when(mockService.createMany(objectContaining(args.input))).thenResolve(output);
        const result = await resolver.createMany({ input: args });
        verify(mockPubSub.publish(anything(), anything())).never();
        return expect(result).toEqual(output);
      });

      it('should not publish an event if enableSubscriptions is true and many.enableSubscriptions is false', async () => {
        const { resolver, mockService, mockPubSub } = await createTestResolver({
          enableSubscriptions: true,
          many: { enableSubscriptions: false },
        });
        const args: CreateManyInputType<Partial<TestResolverDTO>> = {
          input: [
            {
              stringField: 'foo',
            },
          ],
        };
        const output: TestResolverDTO[] = [
          {
            id: 'id-1',
            stringField: 'foo',
          },
        ];
        when(mockService.createMany(objectContaining(args.input))).thenResolve(output);
        const result = await resolver.createMany({ input: args });
        verify(mockPubSub.publish(anything(), anything())).never();
        return expect(result).toEqual(output);
      });
    });

    describe('createSubscription', () => {
      it('should propagate events if enableSubscriptions is true', async () => {
        const { resolver, mockPubSub } = await createTestResolver({
          enableSubscriptions: true,
        });
        const eventName = getDTOEventName(EventType.CREATED, TestResolverDTO);

        const event: CreatedEvent<TestResolverDTO> = {
          [eventName]: {
            id: 'id-1',
            stringField: 'foo',
          },
        };
        const mockIterator = mock<AsyncIterator<CreatedEvent<TestResolverDTO>>>();
        when(mockPubSub.asyncIterator(eventName)).thenReturn(instance(mockIterator));
        when(mockIterator.next()).thenResolve({ done: false, value: event });
        const result = await resolver.createdSubscription().next();
        verify(mockPubSub.asyncIterator(eventName)).once();
        return expect(result).toEqual({
          done: false,
          value: event,
        });
      });

      it('should not propagate events if enableSubscriptions is false', async () => {
        const { resolver } = await createTestResolver({
          enableSubscriptions: false,
        });
        const eventName = getDTOEventName(EventType.CREATED, TestResolverDTO);
        return expect(() => resolver.createdSubscription()).toThrow(`Unable to subscribe to ${eventName}`);
      });
    });
  });
});
