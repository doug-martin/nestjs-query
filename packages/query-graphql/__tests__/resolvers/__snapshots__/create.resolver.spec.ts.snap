// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateResolver #createMany should not create a new type if the CreateManyArgs is supplied 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

input CreateManyInput {
  """Array of records to create"""
  testResolvers: [TestResolverInputDTO!]!
}

input TestResolverInputDTO {
  id: ID!
  stringField: String!
}

`;

exports[`CreateResolver #createMany should not expose create many method if disabled 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

`;

exports[`CreateResolver #createOne should not expose create one method if disabled 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

`;

exports[`CreateResolver #createOne should use the provided CreateOneInput type 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneInput {
  """The record to create"""
  createResolverDTO: TestResolverInputDTO!
}

input TestResolverInputDTO {
  id: ID!
  stringField: String!
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

`;

exports[`CreateResolver created subscription should add subscription types if enableSubscriptions is true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

type Subscription {
  createdTestResolverDTO(input: CreateTestResolverDTOSubscriptionFilterInput): TestResolverDTO!
}

input CreateTestResolverDTOSubscriptionFilterInput {
  """Specify to filter the records returned."""
  filter: TestResolverDTOSubscriptionFilter!
}

input TestResolverDTOSubscriptionFilter {
  and: [TestResolverDTOSubscriptionFilter!]
  or: [TestResolverDTOSubscriptionFilter!]
  id: IDFilterComparison
  stringField: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

`;

exports[`CreateResolver created subscription should not expose subscriptions if enableSubscriptions is false 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

`;

exports[`CreateResolver should create a CreateResolver for the DTO 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

`;

exports[`CreateResolver should not expose create methods if disabled 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`CreateResolver should use the CreateDTOClass if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: TestResolverInputDTO!
}

input TestResolverInputDTO {
  id: ID!
  stringField: String!
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [TestResolverInputDTO!]!
}

`;

exports[`CreateResolver should use the dtoName if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTest(input: CreateOneTestInput!): TestResolverDTO!
  createManyTests(input: CreateManyTestsInput!): [TestResolverDTO!]!
}

input CreateOneTestInput {
  """The record to create"""
  test: CreateTest!
}

input CreateTest {
  id: ID
  stringField: String
}

input CreateManyTestsInput {
  """Array of records to create"""
  tests: [CreateTest!]!
}

`;

exports[`CreateResolver should use the many.name option for the createMany if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  createOneTestResolverDTO(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  create_many_test(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

`;

exports[`CreateResolver should use the one.name option for the createOne if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  create_one_test(input: CreateOneTestResolverDTOInput!): TestResolverDTO!
  createManyTestResolverDTOS(input: CreateManyTestResolverDTOSInput!): [TestResolverDTO!]!
}

input CreateOneTestResolverDTOInput {
  """The record to create"""
  testResolverDTO: CreateTestResolverDTO!
}

input CreateTestResolverDTO {
  id: ID
  stringField: String
}

input CreateManyTestResolverDTOSInput {
  """Array of records to create"""
  testResolverDTOS: [CreateTestResolverDTO!]!
}

`;
